core/lib/state/src/rocksdb/mod.rs

```diff
@@ -131,23 +152,157 @@ impl RocksdbStorage {
+            .into_iter()
+            .partition_map(|(key, (value, index))| match index {
+                Some(index) => Either::Left((key, (value, index))),
+                None => Either::Right((key, value)),
+            });
+        let keys_with_unknown_indices: Vec<_> = logs_with_unknown_indices
+            .iter()
+            .map(|(key, _)| key.hashed_key())
+            .collect();
+
+        let enum_indices = conn
+            .storage_logs_dedup_dal()
+            .enum_indices_for_keys(&keys_with_unknown_indices)
+            .await;
+        assert_eq!(keys_with_unknown_indices.len(), enum_indices.len());
+        self.pending_patch.state = logs_with_known_indices
+            .into_iter()
+            .chain(
+                logs_with_unknown_indices
+                    .into_iter()
+                    .zip(enum_indices)
+                    .map(|((key, value), index)| (key, (value, index))),
+            )
+            .collect();
+    }
+
+    async fn save_missing_enum_indices(&self, conn: &mut StorageProcessor<'_>) {
+        let (Some(start_from), true) = (
+            self.enum_migration_start_from(),
+            self.enum_index_migration_chunk_size > 0,
+        ) else {
+            return;
+        };
+
+        let started_at = Instant::now();
+        let progress_percentage = BigDecimal::from(100)
+            * u256_to_big_decimal(h256_to_u256(start_from))
+            / u256_to_big_decimal(U256::MAX);
+        tracing::info!(
+            "RocksDB enum index migration is not finished, current progress {}%, starting from key {start_from:?}",
+            progress_percentage.with_prec(5)
+        );
+
+        let mut write_batch = self.db.new_write_batch();
+        let (keys, values): (Vec<_>, Vec<_>) = self
+            .db
+            .from_iterator_cf(StateKeeperColumnFamily::State, start_from.as_bytes())
+            .take(self.enum_index_migration_chunk_size)
+            .filter_map(|(key, value)| {
+                (key.as_ref() != Self::BLOCK_NUMBER_KEY
+                    && key.as_ref() != Self::ENUM_INDEX_MIGRATION_PROGRESS_KEY
+                    && key.as_ref() != Self::ENUM_INDEX_MIGRATION_FINISHED
+                    && value.len() != 40)
+                    .then(|| (H256::from_slice(&key), H256::from_slice(&value[..32])))
+            })
+            .unzip();
+        let indices = conn
+            .storage_logs_dedup_dal()
+            .enum_indices_for_keys(&keys)
+            .await;
+        for ((key, value), index) in keys.iter().zip(values).zip(indices) {
+            write_batch.put_cf(
+                StateKeeperColumnFamily::State,
+                key.as_bytes(),
+                &Self::serialize_state_value(value, index),
+            );
+        }
+
+        let next_key = keys
+            .last()
+            .and_then(|last_key| h256_to_u256(*last_key).checked_add(U256::one()))
+            .map(u256_to_h256);
+        match next_key {
+            None => {
```

Comment:
- Nit: `if let Some(_) = next_key` is considered more idiomatic than `match`.
- If you replace `filter_map` and `take` as suggested in the comment above, you would be able to detect the last chunk by the number of keys in it (no need to perform an extra iteration to determine that the next chunk is empty).